analyses/focal-cn-file-preparation/05-define-most-focal-cn-units.Rmd

```diff
@@ -0,0 +1,357 @@
+    toc_float: TRUE
+author: Chante Bethell for ALSF CCDL
+date: 2020
+params:
+  is_ci: FALSE
+---
+
+This notebook defines the most focal recurrent copy number units by removing focal changes that are within entire chromosome arm losses and gains.
+_Most focal_ here meaning:
+
+- If a chromosome arm is not clearly defined as a gain or loss (and is callable) we look to define the cytoband level status
+- If a cytoband is not clearly defined as a gain or loss (and is callable) we then look to define the gene level status
+
+## Usage
+
+This notebook is intended to be run from the command line with the following (assumes you are in the root directory of the repository):
+
+```
+Rscript -e "rmarkdown::render('analyses/focal-cn-file-preparation/05-define-most-focal-cn-units.Rmd', clean = TRUE)"
+```
+
+## Set up
+
+### Libraries and functions
+
+```{r}
+library(tidyverse)
+```
+
+### Custom Function
+
+```{r}
+status_majority_caller <- function(status_df,
+                                   region_variable) {
+  # Given a data.frame with cytoband/gene level copy number status data,
+  # find the dominant status of the cytoband/gene region by calculating
+  # the percentages of the region that each call represents.
+  #
+  # Args:
+  #   status_df: data.frame with cytoband/gene level copy number status data
+  #   region_variable: string of the column name/region to calculate copy number
+  #                    status percentages for
+  # Return:
+  #   status_count: data.frame with percentage values for each unique status in
+  #                 each unique region (arm/cytoband/gene) and the dominant
+  #                 status for that region
+
+  region_sym <- rlang::sym(region_variable)
+
+  if (region_variable == "chromosome_arm") {
+    status_count <- status_df %>%
+      count(!!region_sym, dominant_status) %>%
+      # Spread the data -- each row represents a unique chromosome arm
+      spread(dominant_status, n) %>%
+      replace_na(list(
+        gain = 0,
+        loss = 0,
+        neutral = 0
+      )) %>%
+      group_by(!!region_sym) %>%
+      # The logic here is to define the arm status based on the majority of calls
+      # in the region -- if the number of calls for a specific status is
+      # responsible for more than 50% of the total calls in that region, then it is
+      # used to define the region's status
+      mutate(
+        percent_gain = gain / (gain + loss + neutral),
+        percent_loss = loss / (gain + loss + neutral),
+        percent_neutral = neutral / (gain + loss + neutral),
+        dominant_status = case_when(
+          percent_gain > 0.5 &
+            percent_gain > 1.5 * percent_loss ~ "gain",
+          percent_loss > 0.5 &
+            percent_loss > 1.5 * percent_gain ~ "loss",
+          percent_neutral > 0.5 &
+            percent_neutral > 1.5 * percent_loss &
+            percent_neutral > 1.5 * percent_gain ~ "neutral",
+          TRUE ~ "none"
+        )
+      )
+  } else {
```

Comment:
I would advise making the else more specific, because if for some reason region_variable isn't `chromosome_arm`, `cytoband` or `gene_symbol` this would still run anyway and we don't necessarily want that?